_posts/2015-04-17-Enums-match-mutation-and-moves.md

```diff
@@ -0,0 +1,740 @@
+                                                        ^~~~~~~~~
+enums.rs:357:13: 357:14 warning: value assigned to `shape` is never read, #[warn(unused_assignments)] on by default
+enums.rs:357             shape = Queen
+                         ^~~~~
+```
+
+This warning is again warranted, as the test fails:
+
+```
+test demo_promote_3 ... FAILED
+
+failures:
+
+---- demo_promote_3 stdout ----
+thread 'demo_promote_3' panicked at 'assertion failed: `(left == right)` (left: `Occupied(White(Pawn))`, right: `Occupied(White(Queen))`)', enums.rs:367
+```
+
+### Keeping `ref`s into referenced data
+
+So, what happened here?
+
+The answer is: Even though we have ensured that `Square` is not
+accidentally copied, there is nothing stopping the *contents* of
+`Square` from being copied. So in a `match` arm like
+```
+&mut Square::Occupied(Piece::Black(mut shape)) => { ... }
+```
+if the discriminant value matches that arm, then it will put a
+copy of its shape into `shape`.
+
+This brings us to a very important point about how `match` works in
+Rust: it matches its inputs in-place, but normal bindings like the
+`mut shape` above will move (or copy) the matched substructure into
+fresh locations.
+
+If you want to capture a reference into the matched substructure,
+either to avoid extraneous copying, or to enable modification of the
+original input value, then you need to use a `ref`-pattern.
+
+With this knowledge in hand, let us try again.
+
+In addition, our fourth try simplifies the code slightly by
+dereferencing the argument `*s`.
+
+```rust
+fn promote_to_queen_4(s: &mut Square) {
+    match *s {
+        Square::Empty => {}
+        Square::Occupied(Piece::White(ref mut shape)) |
+        Square::Occupied(Piece::Black(ref mut shape)) => {
+            *shape = Queen
+        }
+    }
+}
+
+#[test]
+fn demo_promote_4() {
+    let mut p = Square::Occupied(Piece::White(Pawn));
+    promote_to_queen_4(&mut p);
+    assert_eq!(p, Square::Occupied(Piece::White(Queen)));
+}
+```
+
+Compiling and running our code now yields:
+
+```
+running 7 tests
+test demo_debug_format ... ok
+test demo_one_letter_collect ... ok
+test demo_one_letter ... ok
+test demo_debug_format_2 ... ok
+test demo_promote_4 ... ok
+test check_renders ... ok
+test demo_promote_functional ... ok
+
+test result: ok. 7 passed; 0 failed; 0 ignored; 0 measured
+```
+
+Huzzah, it works!
+
+The dereference `*s` works because `match` works on both "L-value" and
+"R-value" expressions; L-value expressions evaluate to memory
+locations, while R-value expressions evaluate to values that are then
+stored, if necessary, in temporary memory locations. It simplified our
+code because allowed us to sidestep writing `&mut` at the beginning of
+each of our patterns, which was not such a great burden for this tiny
+code snippet, but can be much more annoying when working with `match`
+expressions that have many more arms, amplifying the redundancy.
+
+## Avoiding unsoundness
```

Comment:
Maybe you could move this into a gist, and just link to it from the post as a kind of appendix?